### PR TITLE
coq-modules: avoid top-level `with ...;` expressions

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). This is a pure refactor: there is no functional change contained in this PR.

Several of the packages asked for `recurseIntoAttrs` to be injected, but didn't actually use it. I removed those.

There are no rebuilds when I run `nixpkgs-review rev HEAD`, so I've targeted master with this PR.

- #208242

## Things done

- [x] Tested compilation of all packages that depend on this change using `nixpkgs-review rev HEAD`.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by fricklerhandwerk — APPROVED** [review]

File: pkgs/development/coq-modules/mathcomp-analysis/default.nix [hunk 1]
```diff
@@ -28,7 +45,7 @@ let
   release."0.3.1".sha256 = "1iad288yvrjv8ahl9v18vfblgqb1l5z6ax644w49w9hwxs93f2k8";
   release."0.2.3".sha256 = "0p9mr8g1qma6h10qf7014dv98ln90dfkwn76ynagpww7qap8s966";
 
-  defaultVersion = with versions; lib.switch [ coq.version mathcomp.version ]  [
```
[the remark this answers is not in the document]
Comment (philiptaron) [reply pkgs/development/coq-modules/mathcomp-analysis/default.nix:31]: I hope so. Note that this `lib` is the `extra-lib` -- it's `lib` with `coq` extras.

File: pkgs/development/coq-modules/metacoq/default.nix [hunk 2]
```diff
@@ -1,11 +1,28 @@
 { lib, fetchzip,
   mkCoqDerivation, recurseIntoAttrs,  single ? false,
   coqPackages, coq, equations, version ? null }@args:
-with builtins // lib;
```
[the remark this answers is not in the document]
Comment (philiptaron) [reply pkgs/development/coq-modules/metacoq/default.nix:4]: # 😬 
Hey, some folks believe in merging attrsets, and this is _definitely_ one way to do it.

**Comment by vbgl** [comment]
Please change the manual first.

**Comment by philiptaron** [comment]
> Please change the manual first.

I did three weeks ago. https://github.com/NixOS/nixpkgs/pull/293829/commits/9f8a58026b4a82bdaf24cdeffcbf8e98ce354984

**Comment by vbgl** [comment]
Here we go: #299927

**Comment by philiptaron** [comment]
> Here we go: #299927

This was merged; what would you like to do next?

**Review by fricklerhandwerk — APPROVED** [review]

**Review by JohnRTitor — COMMENTED** [review]
Probably a good idea to explicitly use `lib.maintainers` and so on, unless we need them multiple times and frequently. Only in those cases, I'd use `inherit`.

**Comment by philiptaron** [comment]
> Probably a good idea to explicitly use `lib.maintainers` and so on, unless we need them multiple times and frequently. Only in those cases, I'd use `inherit`.

I took a closer, hand-crafted look at these patches and updated them to use `lib.` where that was needed, instead of using `inherit (lib) ...`. The result looks OK to my eyes. 

Several of the packages asked for `recurseIntoAttrs` to be injected, but didn't actually use it. I removed those.

I didn't style the code at all beyond what was necessary.

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 299010` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Review by JohnRTitor — APPROVED** [review]
LGTM

**Pull request merged**